File: internal/workplan/workplan.go
// Package workplan owns the dated daily notes.
//
// One note per day lives in a reserved folder, named for its date. When a new
// day starts, unfinished items roll forward from the most recent previous day
// and completed ones are left behind on the day they were completed — that
// carry-forward is the behaviour the whole application exists for.
//
// Everything here is idempotent. Ensure runs on launch, on a midnight timer, and
// whenever the window regains focus, so running it repeatedly must never
// duplicate an item or disturb a day the user has already edited.
package workplan

import (
	"crypto/rand"
	"errors"
	"fmt"
	"io/fs"
	"regexp"
	"strings"
	"time"

	"github.com/oklog/ulid/v2"

	"github.com/vishnu-kyatannawar/nota/internal/mdnote"
	"github.com/vishnu-kyatannawar/nota/internal/vault"
)

// DateLayout is how a workplan note is named and dated.
const DateLayout = "2006-01-02"

// TemplatePath is the vault-relative file holding the recurring items.
const TemplatePath = ".nota/templates/recurring.md"

// Day types. Anything other than work pins the day's hours at zero by default,
// though the user may still log time against it.
const (
	DayWork    = "work"
	DayWeekend = "weekend"
	DayLeave   = "leave"
	DayHoliday = "holiday"
)

var validDayTypes = map[string]bool{
	DayWork: true, DayWeekend: true, DayLeave: true, DayHoliday: true,
}

// ErrInvalidHours is returned for a duration that is not "hh:mm".
var ErrInvalidHours = errors.New("hours must be in hh:mm form, for example 07:30")

// ErrInvalidDayType is returned for a day type outside the known set.
var ErrInvalidDayType = errors.New("unknown day type")

// Options configures the manager.
type Options struct {
	// Folder is the vault-relative reserved folder for daily notes.
	Folder string
	// CreateOnWeekends decides whether Saturday and Sunday get a note at all.
	CreateOnWeekends bool
	// NewID mints ids for seeded items. Injectable so tests can pin them down;
	// defaults to a ULID, which sorts by creation time.
	NewID func() string
}

// Manager creates and rolls over the daily notes.
type Manager struct {
	vault *vault.Vault
	opts  Options
}

// New returns a manager over a vault.
func New(v *vault.Vault, opts Options) *Manager {
	if opts.Folder == "" {
		opts.Folder = "Workplans"
	}
	if opts.NewID == nil {
		opts.NewID = func() string {
			return ulid.MustNew(ulid.Timestamp(time.Now()), rand.Reader).String()
		}
	}
	return &Manager{vault: v, opts: opts}
}

// PathFor is the vault-relative path of a given day's note.
func (m *Manager) PathFor(day time.Time) string {
	return m.opts.Folder + "/" + day.Format(DateLayout) + mdnote.Ext
}

// Ensure makes sure the note for day exists, creating it by rolling the previous
// workplan forward and seeding any recurring items that are due.
//
// It returns the note's path, or an empty path when the day is a weekend and
// weekend notes are switched off. An existing note is never touched, so a day the
// user has already worked on is safe from every later call.
func (m *Manager) Ensure(day time.Time) (string, error) {
	if !m.opts.CreateOnWeekends && isWeekend(day) {
		return "", nil
	}

	path := m.PathFor(day)
	if _, err := m.vault.ReadNote(path); err == nil {
		return path, nil
	}

	note := &mdnote.Note{
		Type:           "workplan",
		Date:           day.Format(DateLayout),
		Hours:          mdnote.FormatDuration(0),
		DayType:        dayTypeFor(day),
		HadFrontmatter: true,
	}

	carried, err := m.carryForward(day)
	if err != nil {
		return "", err
	}
	note.Items = carried

	seeded, err := m.seedRecurring(day, note.Items)
	if err != nil {
		return "", err
	}
	note.Items = append(note.Items, seeded...)

	if err := m.vault.WriteNote(path, note); err != nil {
		return "", err
	}
	return path, nil
}

// carryForward reads the most recent earlier workplan and returns its unfinished
// items, ready to be written into the new day.
func (m *Manager) carryForward(day time.Time) ([]mdnote.Item, error) {
	prevPath, prevDate, err := m.previousWorkplan(day)
	if err != nil || prevPath == "" {
		return nil, err
	}

	prev, err := m.vault.ReadNote(prevPath)
	if err != nil {
		return nil, err
	}

	var out []mdnote.Item
	for _, it := range prev.Items {
		if it.Done {
			// Completed work stays on the day it was completed.
			continue
		}
		// The item keeps its identity, its original creation time and anything
		// logged against it; only the completion state and carry counters move.
		it.DoneAt = ""
		if it.From == "" {
			it.From = prevDate
		}
		it.Carried++
		out = append(out, it)
	}
	return out, nil
}

// previousWorkplan finds the latest dated note before day. It deliberately looks
// for the most recent workplan rather than literally yesterday, so weekends,
// leave and holidays do not break the chain.
func (m *Manager) previousWorkplan(day time.Time) (path, date string, err error) {
	paths, err := m.vault.ListNotes()
	if err != nil {
		return "", "", err
	}
	prefix := m.opts.Folder + "/"
	today := day.Format(DateLayout)

	best := ""
	for _, p := range paths {
		if !strings.HasPrefix(p, prefix) {
			continue
		}
		name := strings.TrimSuffix(strings.TrimPrefix(p, prefix), mdnote.Ext)
		if _, perr := time.Parse(DateLayout, name); perr != nil {
			continue
		}
		if name >= today {
			continue
		}
		if name > best {
			best, path = name, p
		}
	}
	return path, best, nil
}

// seedRecurring returns the recurring items due on day that are not already
// present. Matching on the recurring id means an item carried over from
// yesterday suppresses today's seed, and one the user deleted for today stays
// deleted for today.
func (m *Manager) seedRecurring(day time.Time, existing []mdnote.Item) ([]mdnote.Item, error) {
	templates, err := m.Templates()
	if err != nil {
		return nil, err
	}

	present := map[string]bool{}
	for _, it := range existing {
		if it.Recurring != "" {
			present[it.Recurring] = true
		}
	}

	// Seeded items are stamped with the clock time Ensure ran at, which for the
	// usual midnight rollover is the start of the day.
	stamp := day.Format("15:04")

	var out []mdnote.Item
	for _, tpl := range templates {
		if present[tpl.ID] || !tpl.DueOn(day) {
			continue
		}
		out = append(out, mdnote.Item{
			// A seeded item needs its own id like any other: without one the
			// editor cannot address it to tick, retext or delete it.
			ID:        m.opts.NewID(),
			Text:      tpl.Text,
			CreatedAt: stamp,
			Recurring: tpl.ID,
		})
	}
	return out, nil
}

// Template is one recurring item and how often it appears.
type Template struct {
	ID      string `json:"id"`
	Text    string `json:"text"`
	Cadence string `json:"cadence"`
}

var cadenceToken = regexp.MustCompile(`\s*@(daily|weekdays|weekly:(?:mon|tue|wed|thu|fri|sat|sun))\b`)

var weekdayNames = map[string]time.Weekday{
	"mon": time.Monday, "tue": time.Tuesday, "wed": time.Wednesday,
	"thu": time.Thursday, "fri": time.Friday, "sat": time.Saturday, "sun": time.Sunday,
}

// DueOn reports whether the template should appear on a given day.
func (t Template) DueOn(day time.Time) bool {
	switch {
	case t.Cadence == "daily":
		return true
	case t.Cadence == "weekdays":
		return !isWeekend(day)
	case strings.HasPrefix(t.Cadence, "weekly:"):
		want, ok := weekdayNames[strings.TrimPrefix(t.Cadence, "weekly:")]
		return ok && day.Weekday() == want
	default:
		return false
	}
}

// Templates reads the recurring items file. A missing file simply means none are
// configured. Each template needs a stable id so seeding can tell whether today
// already has it; one is derived from the text and written back on first read, so
// the file stays hand-editable without the user having to invent ids.
func (m *Manager) Templates() ([]Template, error) {
	raw, err := m.vault.ReadRaw(TemplatePath)
	if errors.Is(err, fs.ErrNotExist) {
		// No template file simply means no recurring items are configured.
		return nil, nil
	}
	if err != nil {
		return nil, fmt.Errorf("reading recurring templates: %w", err)
	}

	note, err := mdnote.Parse(raw)
	if err != nil {
		return nil, fmt.Errorf("parsing recurring templates: %w", err)
	}

	var (
		out     []Template
		changed bool
	)
	for i := range note.Items {
		it := &note.Items[i]

		cadence := "daily"
		if m := cadenceToken.FindStringSubmatch(it.Text); m != nil {
			cadence = m[1]
		}
		text := strings.TrimSpace(cadenceToken.ReplaceAllString(it.Text, ""))
		if text == "" {
			continue
		}

		if it.Recurring == "" {
			it.Recurring = slug(text)
			changed = true
		}
		out = append(out, Template{ID: it.Recurring, Text: text, Cadence: cadence})
	}

	if changed {
		// Best effort: the ids are a convenience, and a read-only vault should
		// not stop recurring items from working.
		_ = m.vault.WriteNote(TemplatePath, note)
	}
	return out, nil
}

var nonSlug = regexp.MustCompile(`[^a-z0-9]+`)

func slug(s string) string {
	out := nonSlug.ReplaceAllString(strings.ToLower(s), "-")
	out = strings.Trim(out, "-")
	if len(out) > 40 {
		out = out[:40]
	}
	if out == "" {
		out = "item"
	}
	return out
}

// SetHours records the hours worked on a day.
func (m *Manager) SetHours(path, hours string) error {
	if _, ok := mdnote.ParseDuration(hours); !ok {
		return fmt.Errorf("%w: got %q", ErrInvalidHours, hours)
	}
	return m.mutate(path, func(n *mdnote.Note) { n.Hours = hours })
}

// SetDayType marks a day as work, weekend, leave or holiday. Marking it pins the
// default at zero hours but does not erase time already logged, since a person
// may genuinely have worked on a day off.
func (m *Manager) SetDayType(path, dayType string) error {
	if !validDayTypes[dayType] {
		return fmt.Errorf("%w: %q", ErrInvalidDayType, dayType)
	}
	return m.mutate(path, func(n *mdnote.Note) { n.DayType = dayType })
}

// SuggestedMinutes totals the time logged against a day's items. It is only ever
// a suggestion to pre-fill the field: plenty of a working day — meetings, calls,
// helping someone — never lands on an action item, so the day total stays a value
// the user owns.
func (m *Manager) SuggestedMinutes(path string) (int, error) {
	note, err := m.vault.ReadNote(path)
	if err != nil {
		return 0, err
	}
	return note.TotalMinutes(), nil
}

func (m *Manager) mutate(path string, apply func(*mdnote.Note)) error {
	note, err := m.vault.ReadNote(path)
	if err != nil {
		return err
	}
	apply(note)
	return m.vault.WriteNote(path, note)
}

func isWeekend(day time.Time) bool {
	switch day.Weekday() {
	case time.Saturday, time.Sunday:
		return true
	default:
		return false
	}
}

func dayTypeFor(day time.Time) string {
	if isWeekend(day) {
		return DayWeekend
	}
	return DayWork
}

File: src/workplan.cpp
/*
 * SPDX-FileCopyrightText: 2026 Vishnu Kyatannawar
 * SPDX-License-Identifier: MIT
 */

#include "workplan.h"

#include "vault.h"

#include <KLocalizedString>

#include <QDateTime>
#include <QRandomGenerator>
#include <QRegularExpression>
#include <QSet>

using namespace Qt::StringLiterals;

namespace Workplan
{
namespace
{

const QRegularExpression &cadenceToken()
{
    static const QRegularExpression re(uR"(\s*@(daily|weekdays|weekly:(?:mon|tue|wed|thu|fri|sat|sun))\b)"_s);
    return re;
}

bool isWeekend(QDate day)
{
    return day.dayOfWeek() == 6 || day.dayOfWeek() == 7;
}

QString dayTypeFor(QDate day)
{
    return isWeekend(day) ? QString(DayWeekend) : QString(DayWork);
}

QString stripCadence(const QString &text)
{
    return QString(text).remove(cadenceToken()).trimmed();
}

QString slug(const QString &text)
{
    static const QRegularExpression nonSlug(uR"([^a-z0-9]+)"_s);
    QString out = text.toLower();
    out.replace(nonSlug, "-"_L1);
    while (out.startsWith(u'-')) {
        out.remove(0, 1);
    }
    while (out.endsWith(u'-')) {
        out.chop(1);
    }
    if (out.size() > 40) {
        out = out.first(40);
    }
    return out.isEmpty() ? u"item"_s : out;
}

} // namespace

bool isValidDayType(const QString &dayType)
{
    return dayType == DayWork || dayType == DayWeekend || dayType == DayLeave || dayType == DayHoliday;
}

QString newId()
{
    // Crockford base32, as ULID specifies: no I, L, O or U, so an id read aloud
    // or typed by hand cannot become a different one.
    static constexpr QLatin1StringView alphabet{"0123456789ABCDEFGHJKMNPQRSTVWXYZ"};

    quint64 timestamp = quint64(QDateTime::currentMSecsSinceEpoch()) & 0xFFFFFFFFFFFFULL;
    QString out(26, u'0');

    // The first 10 characters are the 48-bit millisecond timestamp, most
    // significant first, which is what makes ids sort by creation time.
    for (int i = 9; i >= 0; --i) {
        out[i] = QChar(alphabet[int(timestamp & 0x1F)]);
        timestamp >>= 5;
    }
    for (int i = 10; i < 26; ++i) {
        out[i] = QChar(alphabet[int(QRandomGenerator::global()->bounded(32))]);
    }
    return out;
}

bool Template::dueOn(QDate day) const
{
    if (cadence == "daily"_L1) {
        return true;
    }
    if (cadence == "weekdays"_L1) {
        return !isWeekend(day);
    }
    if (cadence.startsWith("weekly:"_L1)) {
        static const QHash<QString, int> weekdays = {
            {u"mon"_s, 1}, {u"tue"_s, 2}, {u"wed"_s, 3}, {u"thu"_s, 4},
            {u"fri"_s, 5}, {u"sat"_s, 6}, {u"sun"_s, 7},
        };
        const auto found = weekdays.constFind(cadence.sliced(7));
        return found != weekdays.cend() && day.dayOfWeek() == *found;
    }
    return false;
}

Manager::Manager(Vault *vault, Options options)
    : m_vault(vault)
    , m_options(std::move(options))
{
    if (m_options.folder.isEmpty()) {
        m_options.folder = u"Workplans"_s;
    }
    if (!m_options.newId) {
        m_options.newId = &Workplan::newId;
    }
}

QString Manager::pathFor(QDate day) const
{
    return m_options.folder + u'/' + day.toString(DateFormat) + MdNote::Ext;
}

QString Manager::ensure(QDate day)
{
    if (!m_options.createOnWeekends && isWeekend(day)) {
        return {};
    }

    const QString path = pathFor(day);
    if (m_vault->exists(path)) {
        // A day the user has already worked on is safe from every later call.
        return path;
    }

    MdNote::Note note;
    note.type = QString(MdNote::TypeWorkplan);
    note.date = day.toString(DateFormat);
    note.hours = MdNote::formatDuration(0);
    note.dayType = dayTypeFor(day);
    note.hadFrontmatter = true;

    note.items = carryForward(day);

    // What repeats sits at the top, above the day's own work.
    QList<MdNote::Item> seeded = seedRecurring(day, note.items);
    seeded.append(note.items);
    note.items = seeded;

    if (!m_vault->writeNote(path, note)) {
        m_lastError = m_vault->lastError();
        return {};
    }
    return path;
}

void Manager::previousWorkplan(QDate day, QString *path, QString *date) const
{
    path->clear();
    date->clear();

    const QString prefix = m_options.folder + u'/';
    const QString today = day.toString(DateFormat);

    // Deliberately the most recent workplan rather than literally yesterday, so
    // weekends, leave and holidays do not break the chain.
    QString best;
    const QStringList paths = m_vault->listNotes();
    for (const QString &p : paths) {
        if (!p.startsWith(prefix)) {
            continue;
        }
        const QString name = p.sliced(prefix.size()).chopped(MdNote::Ext.size());
        if (!QDate::fromString(name, DateFormat).isValid()) {
            continue;
        }
        if (name >= today) {
            continue;
        }
        if (name > best) {
            best = name;
            *path = p;
        }
    }
    *date = best;
}

QList<MdNote::Item> Manager::carryForward(QDate day)
{
    QString prevPath;
    QString prevDate;
    previousWorkplan(day, &prevPath, &prevDate);
    if (prevPath.isEmpty()) {
        return {};
    }

    const auto previous = m_vault->readNote(prevPath);
    if (!previous) {
        return {};
    }

    QList<MdNote::Item> out;
    std::optional<MdNote::Item> pendingHeading;

    for (const MdNote::Item &item : previous->items) {
        if (item.isHeading()) {
            // Held back until something under it turns out to carry, so a group
            // that was entirely finished disappears with its heading.
            pendingHeading = item;
            continue;
        }
        if (!item.recurring.isEmpty()) {
            // A repeating item starts each day fresh, ticked or not. Carrying
            // it would bring yesterday's tick or a growing carry badge with it,
            // and seedRecurring puts a clean copy at the top instead.
            continue;
        }
        if (item.done) {
            // Completed work stays on the day it was completed.
            continue;
        }

        if (pendingHeading) {
            out.append(*pendingHeading);
            pendingHeading.reset();
        }

        // The item keeps its identity, its original creation time and anything
        // logged against it; only the completion state and carry counters move.
        MdNote::Item carried = item;
        carried.doneAt.clear();
        if (carried.from.isEmpty()) {
            carried.from = prevDate;
        }
        ++carried.carried;
        out.append(carried);
    }
    return out;
}

QList<MdNote::Item> Manager::seedRecurring(QDate day, const QList<MdNote::Item> &existing)
{
    const QList<Template> all = templates();

    // Matching on the recurring id means an item carried over from yesterday
    // suppresses today's seed, and one the user deleted for today stays
    // deleted for today.
    QSet<QString> present;
    for (const MdNote::Item &item : existing) {
        if (!item.recurring.isEmpty()) {
            present.insert(item.recurring);
        }
    }

    // Seeded items are stamped with the clock time ensure() ran at, which for
    // the usual midnight rollover is the start of the day.
    const QString stamp = QDateTime(day, QTime::currentTime()).toString(u"HH:mm"_s);

    QList<MdNote::Item> out;
    for (const Template &tpl : all) {
        if (present.contains(tpl.id) || !tpl.dueOn(day)) {
            continue;
        }
        MdNote::Item item;
        // A seeded item needs its own id like any other: without one the editor
        // cannot address it to tick, retext or delete it.
        item.id = m_options.newId();
        item.text = tpl.text;
        item.createdAt = stamp;
        item.recurring = tpl.id;
        out.append(item);
    }
    return out;
}

QList<Template> Manager::templates()
{
    const auto raw = m_vault->readRaw(QString(TemplatePath));
    if (!raw) {
        // No template file simply means no recurring items are configured.
        return {};
    }

    MdNote::Note note = MdNote::parse(*raw);

    QList<Template> out;
    bool changed = false;

    for (MdNote::Item &item : note.items) {
        if (item.isHeading()) {
            continue;
        }
        QString cadence = u"daily"_s;
        const QRegularExpressionMatch m = cadenceToken().match(item.text);
        if (m.hasMatch()) {
            cadence = m.captured(1);
        }
        const QString text = stripCadence(item.text);
        if (text.isEmpty()) {
            continue;
        }
        if (item.recurring.isEmpty()) {
            item.recurring = slug(text);
            changed = true;
        }
        out.append(Template{item.recurring, text, cadence});
    }

    if (changed) {
        // Best effort: the ids are a convenience, and a read-only vault should
        // not stop recurring items from working.
        m_vault->writeNote(QString(TemplatePath), note);
    }
    return out;
}

std::optional<MdNote::Note> Manager::templateNote()
{
    // templates() runs first because it writes back any missing ids, which is
    // what lets every mutation below address a line by a stable id.
    templates();

    const auto raw = m_vault->readRaw(QString(TemplatePath));
    if (!raw) {
        return MdNote::Note{};
    }
    return MdNote::parse(*raw);
}

bool Manager::addTemplate(const QString &rawText, Template *added)
{
    const QString text = stripCadence(rawText);
    if (text.isEmpty()) {
        m_lastError = i18n("A repeating item needs some text.");
        return false;
    }
    auto note = templateNote();
    if (!note) {
        return false;
    }

    Template tpl{m_options.newId(), text, u"daily"_s};

    MdNote::Item item;
    item.text = text;
    item.recurring = tpl.id;
    note->items.append(item);

    if (!m_vault->writeNote(QString(TemplatePath), *note)) {
        m_lastError = m_vault->lastError();
        return false;
    }
    if (added) {
        *added = tpl;
    }
    return true;
}

bool Manager::removeTemplate(const QString &id)
{
    auto note = templateNote();
    if (!note) {
        return false;
    }

    QList<MdNote::Item> kept;
    kept.reserve(note->items.size());
    for (const MdNote::Item &item : std::as_const(note->items)) {
        if (item.recurring == id) {
            continue;
        }
        kept.append(item);
    }
    if (kept.size() == note->items.size()) {
        return true;
    }
    note->items = kept;

    if (!m_vault->writeNote(QString(TemplatePath), *note)) {
        m_lastError = m_vault->lastError();
        return false;
    }
    return true;
}

bool Manager::renameTemplate(const QString &id, const QString &rawText)
{
    const QString text = stripCadence(rawText);
    if (text.isEmpty()) {
        m_lastError = i18n("A repeating item needs some text.");
        return false;
    }
    auto note = templateNote();
    if (!note) {
        return false;
    }

    bool changed = false;
    for (MdNote::Item &item : note->items) {
        if (item.recurring != id) {
            continue;
        }
        // The cadence lives in the text as an @token; keep whatever was there.
        const QString cadence = cadenceToken().match(item.text).captured(0);
        const QString next = text + cadence;
        if (next != item.text) {
            item.text = next;
            changed = true;
        }
    }
    if (!changed) {
        return true;
    }

    if (!m_vault->writeNote(QString(TemplatePath), *note)) {
        m_lastError = m_vault->lastError();
        return false;
    }
    return true;
}

bool Manager::seedInto(QDate day)
{
    const QString path = pathFor(day);
    auto note = m_vault->readNote(path);
    if (!note) {
        m_lastError = m_vault->lastError();
        return false;
    }

    QList<MdNote::Item> seeded = seedRecurring(day, note->items);
    if (seeded.isEmpty()) {
        return true;
    }
    seeded.append(note->items);
    note->items = seeded;

    if (!m_vault->writeNote(path, *note)) {
        m_lastError = m_vault->lastError();
        return false;
    }
    return true;
}

bool Manager::dropFrom(QDate day, const QString &id)
{
    const QString path = pathFor(day);
    if (!m_vault->exists(path)) {
        return true;
    }
    auto note = m_vault->readNote(path);
    if (!note) {
        m_lastError = m_vault->lastError();
        return false;
    }

    QList<MdNote::Item> kept;
    kept.reserve(note->items.size());
    for (const MdNote::Item &item : std::as_const(note->items)) {
        if (item.recurring == id) {
            continue;
        }
        kept.append(item);
    }
    if (kept.size() == note->items.size()) {
        return true;
    }
    note->items = kept;

    if (!m_vault->writeNote(path, *note)) {
        m_lastError = m_vault->lastError();
        return false;
    }
    return true;
}

bool Manager::mutate(const QString &path, const std::function<void(MdNote::Note &)> &apply)
{
    auto note = m_vault->readNote(path);
    if (!note) {
        m_lastError = m_vault->lastError();
        return false;
    }
    apply(*note);
    if (!m_vault->writeNote(path, *note)) {
        m_lastError = m_vault->lastError();
        return false;
    }
    return true;
}

bool Manager::setHours(const QString &path, const QString &hours)
{
    if (!MdNote::parseDuration(hours, nullptr)) {
        m_lastError = i18n("Hours must be written as hh:mm, for example 07:30.");
        return false;
    }
    return mutate(path, [&hours](MdNote::Note &note) {
        note.hours = hours;
    });
}

bool Manager::setDayType(const QString &path, const QString &dayType)
{
    if (!isValidDayType(dayType)) {
        m_lastError = i18n("Unknown day type: %1", dayType);
        return false;
    }
    return mutate(path, [&dayType](MdNote::Note &note) {
        note.dayType = dayType;
    });
}

int Manager::suggestedMinutes(const QString &path) const
{
    const auto note = m_vault->readNote(path);
    return note ? note->totalMinutes() : 0;
}

} // namespace Workplan
